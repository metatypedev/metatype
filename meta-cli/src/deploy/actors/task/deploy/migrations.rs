// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use color_eyre::owo_colors::OwoColorize;

use super::{DeployAction, DeployActionInner, Migration, MigrationActionOverride};
use crate::deploy::actors::console::input::{Confirm, Select};
use crate::deploy::actors::console::Console;
use crate::deploy::actors::task::action::ActionFinalizeContext;
use crate::deploy::actors::task::TaskActor;
use crate::interlude::*;

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
#[allow(unused)]
pub struct DatabaseResetRequired {
    message: String,
    runtime_name: String,
}

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
#[allow(unused)]
pub struct NullConstraintViolation {
    message: String,
    runtime_name: String,
    column: String,
    migration_name: String,
    is_new_column: bool,
    table: String,
}

#[allow(unused)]
struct ResolveNullConstraintViolation {
    failure: NullConstraintViolation,
}

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GenericPushFailure {
    message: String,
}

#[derive(Deserialize, Debug, Clone)]
#[serde(tag = "reason")]
enum PushFailure {
    Unknown(GenericPushFailure),
    DatabaseResetRequired(DatabaseResetRequired),
    NullConstraintViolation(NullConstraintViolation),
}

type RuntimeName = String;

impl DeployActionInner {
    // TODO why not async?
    pub(super) fn unpack_migrations(
        &self,
        tg_name: &str,
        migrations: &[Migration],
        ctx: &ActionFinalizeContext<Arc<Self>>,
        scope: &impl std::fmt::Display,
    ) {
        let migdir = ctx.config.prisma_migration_dir_abs(tg_name);

        for migration in migrations.iter() {
            let dest = migdir.join(&migration.runtime);
            if let Err(err) = common::archive::unpack(&dest, Some(&migration.archive)) {
                ctx.console.error(format!(
                    "{scope} error while unpacking migrations into {:?}",
                    migdir
                ));
                ctx.console.error(format!("{err:?}"));
            } else {
                ctx.console.info(format!(
                    "{scope} unpacked migrations for {}/{} at {}",
                    tg_name.cyan(),
                    migration.runtime,
                    dest.display().bold()
                ));
            }
        }
    }

    pub(super) async fn handle_push_failure(
        &self,
        tg_name: &str,
        typegraph_path: &Path,
        failure_raw: &str,
        ctx: &ActionFinalizeContext<Arc<Self>>,
        scope: &impl std::fmt::Display,
    ) -> Result<Option<(RuntimeName, MigrationActionOverride)>> {
        let failure = serde_json::from_str::<PushFailure>(failure_raw)
            .context("failed to parse failure data")?;
        match failure {
            PushFailure::Unknown(error) => {
                ctx.console.error(format!(
                    "{scope} unknown error: {msg}",
                    scope = scope,
                    msg = error.message,
                ));

                Ok(None)
            }

            PushFailure::DatabaseResetRequired(error) => {
                ctx.console
                    .error(format!("{scope} {message}", message = error.message));
                ctx.console.warning(format!(
                    "{scope} database reset required for prisma runtime {rt} in typegraph {name}",
                    name = tg_name.cyan(),
                    rt = error.runtime_name.magenta(),
                ));

                let reset = Confirm::new(
                    ctx.console.clone(),
                    format!(
                        "{scope} Do you want to reset the database for prisma runtime {rt} in typegraph {name}?",
                        scope = scope.yellow(),
                        name = tg_name.cyan(),
                        rt = error.runtime_name.magenta(),
                    ),
                ).interact( ).await.context("failed to read user input")?;

                if reset {
                    Ok(Some((
                        error.runtime_name,
                        MigrationActionOverride::ResetDatabase,
                    )))
                } else {
                    Ok(None)
                }
            }

            PushFailure::NullConstraintViolation(error) => {
                ctx.console
                    .error(format!("{scope} {message}", message = error.message));

                if error.is_new_column {
                    ctx.console.info(format!("{scope} manually edit the migration {migration} or remove the migration and set a default value in the typegraph", migration = error.migration_name));
                }

                use options::ConstraintViolationOptions as Options;
                let (_, choice) = Select::new(
                    ctx.console.clone(),
                    "Choose one of the following options".to_string(),
                )
                .interact(&[
                    Box::new(options::RemoveLatestMigration),
                    Box::new(options::ManualResolution {
                        message: Some(format!(
                            "Set a default value for the column `{}` in the table `{}`",
                            error.column, error.table
                        )),
                    }),
                    Box::new(options::ForceReset),
                ])
                .await
                .context("failed to read user input: {err}")?;

                match choice {
                    Options::RemoveLatestMigration => {
                        let migration_path = ctx
                            .config
                            .prisma_migration_dir_abs(tg_name)
                            .join(&error.runtime_name)
                            .join(&error.migration_name);
                        tokio::fs::remove_dir_all(&migration_path)
                            .await
                            .with_context(|| {
                                format!("failed to remove migrations at {migration_path:?}")
                            })?;
                        ctx.console
                            .info(format!("Removed migration directory: {migration_path:?}"));
                        ctx.console.info(format!("You can now update your typegraph at {} to create an alternative non-breaking schema.", typegraph_path.to_str().unwrap().bold()));
                        Ok(Some((
                            error.runtime_name,
                            MigrationActionOverride::ResetDatabase,
                        )))
                    }
                    Options::ManualResolution => {
                        let migration_path = ctx
                            .config
                            .prisma_migration_dir_abs(tg_name)
                            .join(&error.runtime_name)
                            .join(&error.migration_name);
                        eprintln!("Edit the migration file at {migration_path:?} then press enter to continue...");

                        ctx.console.read_line().await;
                        Ok(Some((
                            error.runtime_name,
                            MigrationActionOverride::ResetDatabase,
                        )))
                    }
                    Options::ForceReset => Ok(Some((
                        error.runtime_name,
                        MigrationActionOverride::ResetDatabase,
                    ))),
                }
            }
        }
    }
}

#[derive(Debug)]
pub struct ConfirmDatabaseResetRequired {
    pub task: Addr<TaskActor<DeployAction>>,
    pub tg_name: String,
    pub runtime_name: String,
}

mod options {
    use crate::deploy::actors::console::input::{OptionLabel, SelectOption};

    pub enum ConstraintViolationOptions {
        RemoveLatestMigration,
        ManualResolution,
        ForceReset,
    }

    #[derive(Debug)]
    pub struct RemoveLatestMigration;

    impl SelectOption<ConstraintViolationOptions> for RemoveLatestMigration {
        fn get_value(&self) -> ConstraintViolationOptions {
            ConstraintViolationOptions::RemoveLatestMigration
        }

        fn label(&self) -> OptionLabel<'_> {
            OptionLabel::new("Remove the latest migration.")
        }
    }

    #[derive(Debug)]
    pub struct ManualResolution {
        pub message: Option<String>,
    }

    impl SelectOption<ConstraintViolationOptions> for ManualResolution {
        fn get_value(&self) -> ConstraintViolationOptions {
            ConstraintViolationOptions::ManualResolution
        }

        fn label(&self) -> OptionLabel<'_> {
            let label = OptionLabel::new("Manually resolve the migration.");
            if let Some(message) = &self.message {
                label.with_secondary(format!("Edit the migration file: {}.", message))
            } else {
                label
            }
        }
    }

    #[derive(Debug)]
    pub struct ForceReset;

    impl SelectOption<ConstraintViolationOptions> for ForceReset {
        fn get_value(&self) -> ConstraintViolationOptions {
            ConstraintViolationOptions::ForceReset
        }

        fn label(&self) -> OptionLabel<'_> {
            OptionLabel::new("Force reset the development database.").with_secondary(
                "Warning: The failed migration will potentially fail again in deployment.",
            )
        }
    }
}
