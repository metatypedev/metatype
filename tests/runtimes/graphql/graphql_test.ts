// Copyright Metatype OÃœ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

import { QueryEngine } from "../../../src/engine/query_engine.ts";
import { removeMigrations } from "../../utils/migrations.ts";
import { dropSchemas, recreateMigrations } from "../../utils/migrations.ts";
import { gql, Meta } from "../../utils/mod.ts";

const PYTHON_TG_PATH = "runtimes/graphql/typegraphs/python/graphql.py";
const TS_TG_PATH = "runtimes/graphql/typegraphs/deno/graphql.ts";

Meta.test("Typegraph generation with GraphQL runtime", async (t) => {
  await t.assertSameTypegraphs(TS_TG_PATH, PYTHON_TG_PATH);
});

async function testEngine(engine: QueryEngine) {
  // get users
  await gql`
    {
      users {
        data {
          id
          name
        }
      }
    }
  `
    .expectData({
      users: {
        data: [
          {
            id: "1",
            name: "Leanne Graham",
          },
          {
            id: "2",
            name: "Ervin Howell",
          },
          {
            id: "3",
            name: "Clementine Bauch",
          },
          {
            id: "4",
            name: "Patricia Lebsack",
          },
          {
            id: "5",
            name: "Chelsey Dietrich",
          },
          {
            id: "6",
            name: "Mrs. Dennis Schulist",
          },
          {
            id: "7",
            name: "Kurtis Weissnat",
          },
          {
            id: "8",
            name: "Nicholas Runolfsdottir V",
          },
          {
            id: "9",
            name: "Glenna Reichert",
          },
          {
            id: "10",
            name: "Clementina DuBuque",
          },
        ],
      },
    })
    .on(engine);

  // get user
  await gql`
    {
      user(id: "1") {
        id
        name
      }
    }
  `
    .expectData({
      user: {
        id: "1",
        name: "Leanne Graham",
      },
    })
    .on(engine);

  // Mutation
  await gql`
    mutation {
      create_message(data: { title: "Hey", user_id: "1" }) {
        id
        title
        user_id
      }
    }
  `
    .expectData({
      create_message: {
        id: 1,
        title: "Hey",
        user_id: "1",
      },
    })
    .on(engine);

  // TODO: Fails because metatype type system is not recognized by outside APIs
  await gql`
    mutation CreateUser($name: String!, $username: String!, $email: String!) {
      createUser(input: { name: $name, username: $username, email: $email }) {
        id
        name
      }
    }
  `
    .withVars({
      name: "John",
      username: "Johhn",
      email: "John@gmail.com",
    })
    .expectData({
      createUser: {
        id: "11",
        name: "John",
      },
    })
    .on(engine);

  await gql`
    {
      messages(take: 4) {
        id
        title
        user_id
        user {
          id
          name
        }
      }
    }
  `
    .expectData({
      messages: [
        {
          id: 1,
          title: "Hey",
          user_id: "1",
          user: {
            id: "1",
            name: "Leanne Graham",
          },
        },
      ],
    })
    .on(engine);
}

Meta.test(
  {
    name: "GraphQL Runtime: Python SDK",
  },
  async (t) => {
    const schema = "graphql-python";
    const connStr =
      `postgresql://postgres:password@localhost:5432/db?schema=${schema}`;

    const engine = await t.engine(PYTHON_TG_PATH, {
      secrets: {
        POSTGRES: connStr,
      },
    });
    await dropSchemas(engine);
    await recreateMigrations(engine);

    await t.should(
      "work when fetching data using GraphQL Runtime",
      async () => {
        await testEngine(engine);
      },
    );
    await removeMigrations(engine);
  },
);

Meta.test(
  {
    name: "GraphQL Runtime: TS SDK",
  },
  async (t) => {
    const schema = "graphql-ts";
    const connStr =
      `postgresql://postgres:password@localhost:5432/db?schema=${schema}`;
    const engine = await t.engine(TS_TG_PATH, {
      secrets: {
        POSTGRES: connStr,
      },
    });
    await dropSchemas(engine);
    await recreateMigrations(engine);

    await t.should(
      "work when fetching data through graphql request",
      async () => {
        await testEngine(engine);
      },
    );
    await removeMigrations(engine);
  },
);
