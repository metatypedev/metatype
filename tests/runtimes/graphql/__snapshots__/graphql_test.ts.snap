export const snapshot = {};

snapshot[`Typegraph generation with GraphQL runtime 1`] = `
'{
  "types": [
    {
      "type": "object",
      "title": "graphql",
      "runtime": 0,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "user": 1,
        "users": 7,
        "createUser": 11,
        "create_message": 14,
        "messages": 27
      },
      "required": [
        "user",
        "users",
        "createUser",
        "create_message",
        "messages"
      ]
    },
    {
      "type": "function",
      "title": "root_user_fn",
      "runtime": 1,
      "policies": [
        0
      ],
      "config": {},
      "as_id": false,
      "input": 2,
      "output": 4,
      "materializer": 0,
      "rate_weight": null,
      "rate_calls": false
    },
    {
      "type": "object",
      "title": "root_user_fn_input",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "id": 3
      },
      "required": []
    },
    {
      "type": "string",
      "title": "root_user_fn_input_id_string",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": true
    },
    {
      "type": "object",
      "title": "User",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "id": 5,
        "name": 6
      },
      "required": []
    },
    {
      "type": "string",
      "title": "User_id_string",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": true
    },
    {
      "type": "string",
      "title": "User_name_string",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": false
    },
    {
      "type": "function",
      "title": "root_users_fn",
      "runtime": 1,
      "policies": [
        0
      ],
      "config": {},
      "as_id": false,
      "input": 8,
      "output": 9,
      "materializer": 2,
      "rate_weight": null,
      "rate_calls": false
    },
    {
      "type": "object",
      "title": "root_users_fn_input",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {},
      "required": []
    },
    {
      "type": "object",
      "title": "root_users_fn_output",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "data": 10
      },
      "required": []
    },
    {
      "type": "list",
      "title": "root_users_fn_output_data_User_list",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": false,
      "items": 4
    },
    {
      "type": "function",
      "title": "root_createUser_fn",
      "runtime": 1,
      "policies": [
        0
      ],
      "config": {},
      "as_id": false,
      "input": 12,
      "output": 4,
      "materializer": 3,
      "rate_weight": null,
      "rate_calls": false
    },
    {
      "type": "object",
      "title": "root_createUser_fn_input",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "input": 13
      },
      "required": []
    },
    {
      "type": "object",
      "title": "CreateUserInput",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "name": 6,
        "username": 6,
        "email": 6
      },
      "required": []
    },
    {
      "type": "function",
      "title": "root_create_message_fn",
      "runtime": 2,
      "policies": [
        0
      ],
      "config": {},
      "as_id": false,
      "input": 23,
      "output": 26,
      "materializer": 4,
      "rate_weight": null,
      "rate_calls": false
    },
    {
      "type": "object",
      "title": "message",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "id": 16,
        "title": 17,
        "user_id": 18,
        "user": 19
      },
      "required": []
    },
    {
      "type": "integer",
      "title": "_message_CreateInput_id_integer",
      "runtime": 2,
      "policies": [],
      "config": {
        "auto": true
      },
      "as_id": true
    },
    {
      "type": "string",
      "title": "_message_CreateInput_title_string",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false
    },
    {
      "type": "string",
      "title": "uid",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false
    },
    {
      "type": "function",
      "title": "_messageOutputType_user_fn",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": false,
      "input": 20,
      "output": 22,
      "materializer": 5,
      "rate_weight": null,
      "rate_calls": false
    },
    {
      "type": "object",
      "title": "_messageOutputType_user_fn_input",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "id": 21
      },
      "required": []
    },
    {
      "type": "string",
      "title": "_messageOutputType_user_fn_input_id_string",
      "runtime": 1,
      "policies": [],
      "injection": {
        "source": "parent",
        "data": {
          "value": 18
        }
      },
      "config": {},
      "as_id": true
    },
    {
      "type": "optional",
      "title": "_messageOutputType_user_fn_output",
      "runtime": 1,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 4,
      "default_value": null
    },
    {
      "type": "object",
      "title": "root_create_message_fn_input",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "data": 24
      },
      "required": []
    },
    {
      "type": "object",
      "title": "_message_CreateInput",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "id": 25,
        "title": 17,
        "user_id": 18
      },
      "required": []
    },
    {
      "type": "optional",
      "title": "_message_CreateInput_id__message_CreateInput_id_integer_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 16,
      "default_value": null
    },
    {
      "type": "object",
      "title": "_messageOutputType",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "id": 16,
        "title": 17,
        "user_id": 18,
        "user": 19
      },
      "required": []
    },
    {
      "type": "function",
      "title": "root_messages_fn",
      "runtime": 2,
      "policies": [
        0
      ],
      "config": {},
      "as_id": false,
      "input": 28,
      "output": 81,
      "materializer": 6,
      "rate_weight": null,
      "rate_calls": false
    },
    {
      "type": "object",
      "title": "_message_QueryInput",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "where": 29,
        "orderBy": 62,
        "take": 69,
        "skip": 71,
        "cursor": 73,
        "distinct": 78
      },
      "required": []
    },
    {
      "type": "optional",
      "title": "_message_QueryInput_where_QuerymessageWhereInput_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 30,
      "default_value": null
    },
    {
      "type": "object",
      "title": "QuerymessageWhereInput",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "id": 31,
        "title": 44,
        "user_id": 44,
        "AND": 60,
        "OR": 60,
        "NOT": 29
      },
      "required": []
    },
    {
      "type": "optional",
      "title": "QuerymessageWhereInput_id__integer_filter_c_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 32,
      "default_value": null
    },
    {
      "type": "optional",
      "title": "_integer_filter_c",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 33,
      "default_value": null
    },
    {
      "type": "union",
      "title": "QuerymessageWhereInput_id_union",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "anyOf": [
        34,
        43
      ]
    },
    {
      "type": "either",
      "title": "_integer_filter",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "oneOf": [
        35,
        36,
        37,
        38,
        40,
        42
      ]
    },
    {
      "type": "integer",
      "title": "_integer_filter_t0_integer",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false
    },
    {
      "type": "object",
      "title": "_integer_filter_t1_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "equals": 35
      },
      "required": []
    },
    {
      "type": "object",
      "title": "_integer_filter_t2_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "not": 35
      },
      "required": []
    },
    {
      "type": "object",
      "title": "_integer_filter_t3_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "lt": 39,
        "gt": 39,
        "lte": 39,
        "gte": 39
      },
      "required": []
    },
    {
      "type": "optional",
      "title": "_integer_filter_t3_struct_lt__integer_filter_t0_integer_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 35,
      "default_value": null
    },
    {
      "type": "object",
      "title": "_integer_filter_t4_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "in": 41
      },
      "required": []
    },
    {
      "type": "list",
      "title": "_integer_filter_t4_struct_in__integer_filter_t0_integer_list",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "items": 35
    },
    {
      "type": "object",
      "title": "_integer_filter_t5_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "notIn": 41
      },
      "required": []
    },
    {
      "type": "object",
      "title": "QuerymessageWhereInput_id_union_t1_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "not": 34
      },
      "required": []
    },
    {
      "type": "optional",
      "title": "QuerymessageWhereInput_title__string_filter_c_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 45,
      "default_value": null
    },
    {
      "type": "optional",
      "title": "_string_filter_c",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 46,
      "default_value": null
    },
    {
      "type": "union",
      "title": "QuerymessageWhereInput_title_union",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "anyOf": [
        47,
        59
      ]
    },
    {
      "type": "union",
      "title": "_string_filter",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "anyOf": [
        17,
        48,
        49,
        50,
        52,
        53,
        56,
        57
      ]
    },
    {
      "type": "object",
      "title": "_string_filter_t1_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "equals": 17
      },
      "required": []
    },
    {
      "type": "object",
      "title": "_string_filter_t2_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "not": 17
      },
      "required": []
    },
    {
      "type": "object",
      "title": "_string_filter_t3_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "in": 51
      },
      "required": []
    },
    {
      "type": "list",
      "title": "_string_filter_t3_struct_in__message_CreateInput_title_string_list",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "items": 17
    },
    {
      "type": "object",
      "title": "_string_filter_t4_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "notIn": 51
      },
      "required": []
    },
    {
      "type": "object",
      "title": "_string_filter_t5_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "contains": 17,
        "mode": 54
      },
      "required": []
    },
    {
      "type": "optional",
      "title": "_string_filter_t5_struct_mode__string_filter_t5_struct_mode_string_enum_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 55,
      "default_value": null
    },
    {
      "type": "string",
      "title": "_string_filter_t5_struct_mode_string_enum",
      "runtime": 2,
      "policies": [],
      "enum": [
        "\\\\"insensitive\\\\""
      ],
      "config": {},
      "as_id": false
    },
    {
      "type": "object",
      "title": "_string_filter_t6_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "search": 17
      },
      "required": []
    },
    {
      "type": "object",
      "title": "_string_filter_t7_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "startsWith": 58,
        "endsWith": 58
      },
      "required": []
    },
    {
      "type": "optional",
      "title": "_string_filter_t7_struct_startsWith__message_CreateInput_title_string_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 17,
      "default_value": null
    },
    {
      "type": "object",
      "title": "QuerymessageWhereInput_title_union_t1_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "not": 47
      },
      "required": []
    },
    {
      "type": "optional",
      "title": "QuerymessageWhereInput_AND_QuerymessageWhereInput_AND_QuerymessageWhereInput_list_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 61,
      "default_value": null
    },
    {
      "type": "list",
      "title": "QuerymessageWhereInput_AND_QuerymessageWhereInput_list",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "items": 30
    },
    {
      "type": "optional",
      "title": "_message_QueryInput_orderBy__message_OrderBy_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 63,
      "default_value": null
    },
    {
      "type": "list",
      "title": "_message_OrderBy",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "items": 64
    },
    {
      "type": "object",
      "title": "_message_QueryInput_orderBy_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "id": 65,
        "title": 65,
        "user_id": 65
      },
      "required": []
    },
    {
      "type": "optional",
      "title": "_Sort",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 66,
      "default_value": null
    },
    {
      "type": "union",
      "title": "_message_QueryInput_orderBy_struct_id_union",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "anyOf": [
        67,
        68
      ]
    },
    {
      "type": "object",
      "title": "_message_QueryInput_orderBy_struct_id_union_t0_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "sort": 68
      },
      "required": []
    },
    {
      "type": "string",
      "title": "_SortOrder",
      "runtime": 2,
      "policies": [],
      "enum": [
        "\\\\"asc\\\\"",
        "\\\\"desc\\\\""
      ],
      "config": {},
      "as_id": false
    },
    {
      "type": "optional",
      "title": "_message_QueryInput_take__Take_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 70,
      "default_value": null
    },
    {
      "type": "integer",
      "title": "_Take",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "exclusiveMinimum": 0
    },
    {
      "type": "optional",
      "title": "_message_QueryInput_skip__Skip_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 72,
      "default_value": null
    },
    {
      "type": "integer",
      "title": "_Skip",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "minimum": 0
    },
    {
      "type": "optional",
      "title": "_message_QueryInput_cursor__message_Cursor_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 74,
      "default_value": null
    },
    {
      "type": "union",
      "title": "_message_Cursor",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "anyOf": [
        75,
        76,
        77
      ]
    },
    {
      "type": "object",
      "title": "_message_Cursor_t0_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "id": 16
      },
      "required": []
    },
    {
      "type": "object",
      "title": "_message_Cursor_t1_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "title": 17
      },
      "required": []
    },
    {
      "type": "object",
      "title": "_message_Cursor_t2_struct",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "user_id": 18
      },
      "required": []
    },
    {
      "type": "optional",
      "title": "_message_QueryInput_distinct__KeysOf_message_optional",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "item": 79,
      "default_value": null
    },
    {
      "type": "list",
      "title": "_KeysOf_message",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "items": 80
    },
    {
      "type": "string",
      "title": "_message_QueryInput_distinct_string_enum",
      "runtime": 2,
      "policies": [],
      "enum": [
        "\\\\"id\\\\"",
        "\\\\"title\\\\"",
        "\\\\"user_id\\\\""
      ],
      "config": {},
      "as_id": false
    },
    {
      "type": "list",
      "title": "root_messages_fn_output",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "items": 82
    },
    {
      "type": "object",
      "title": "messageWithNestedCount",
      "runtime": 2,
      "policies": [],
      "config": {},
      "as_id": false,
      "properties": {
        "id": 16,
        "title": 17,
        "user_id": 18,
        "user": 19
      },
      "required": []
    }
  ],
  "materializers": [
    {
      "name": "query",
      "runtime": 1,
      "effect": {
        "effect": "read",
        "idempotent": true
      },
      "data": {
        "path": null
      }
    },
    {
      "name": "predefined_function",
      "runtime": 0,
      "effect": {
        "effect": "read",
        "idempotent": true
      },
      "data": {
        "name": "true"
      }
    },
    {
      "name": "query",
      "runtime": 1,
      "effect": {
        "effect": "read",
        "idempotent": true
      },
      "data": {
        "path": null
      }
    },
    {
      "name": "mutation",
      "runtime": 1,
      "effect": {
        "effect": "create",
        "idempotent": false
      },
      "data": {
        "path": null
      }
    },
    {
      "name": "prisma_operation",
      "runtime": 2,
      "effect": {
        "effect": "create",
        "idempotent": false
      },
      "data": {
        "table": "message",
        "operation": "createOne"
      }
    },
    {
      "name": "query",
      "runtime": 1,
      "effect": {
        "effect": "read",
        "idempotent": true
      },
      "data": {
        "path": null
      }
    },
    {
      "name": "prisma_operation",
      "runtime": 2,
      "effect": {
        "effect": "read",
        "idempotent": true
      },
      "data": {
        "table": "message",
        "operation": "findMany"
      }
    }
  ],
  "runtimes": [
    {
      "name": "deno",
      "data": {
        "worker": "default",
        "permissions": {}
      }
    },
    {
      "name": "graphql",
      "data": {
        "endpoint": "https://graphqlzero.almansi.me/api"
      }
    },
    {
      "name": "prisma",
      "data": {
        "name": "graphql",
        "connection_string_secret": "POSTGRES",
        "models": [
          {
            "typeIdx": 15,
            "typeName": "message",
            "props": [
              {
                "type": "scalar",
                "key": "id",
                "propType": {
                  "type": "Int"
                },
                "cardinality": "one",
                "typeIdx": 16,
                "injection": null,
                "unique": false,
                "auto": true,
                "defaultValue": null
              },
              {
                "type": "scalar",
                "key": "title",
                "propType": {
                  "type": "String",
                  "format": "Plain"
                },
                "cardinality": "one",
                "typeIdx": 17,
                "injection": null,
                "unique": false,
                "auto": false,
                "defaultValue": null
              },
              {
                "type": "scalar",
                "key": "user_id",
                "propType": {
                  "type": "String",
                  "format": "Plain"
                },
                "cardinality": "one",
                "typeIdx": 18,
                "injection": null,
                "unique": false,
                "auto": false,
                "defaultValue": null
              }
            ],
            "idFields": [
              "id"
            ],
            "uniqueConstraints": []
          }
        ],
        "relationships": [],
        "migration_options": {
          "migration_files": null,
          "create": false,
          "reset": false
        }
      }
    }
  ],
  "policies": [
    {
      "name": "__public",
      "materializer": 1
    }
  ],
  "meta": {
    "prefix": null,
    "secrets": [],
    "queries": {
      "dynamic": true,
      "endpoints": []
    },
    "cors": {
      "allow_origin": [],
      "allow_headers": [],
      "expose_headers": [],
      "allow_methods": [],
      "allow_credentials": true,
      "max_age_sec": null
    },
    "auths": [],
    "rate": null,
    "version": "0.0.3",
    "randomSeed": null,
    "artifacts": {}
  }
}'
`;
